### PR TITLE
Refactor pagination SQL queries to use generic paramaters

### PR DESCRIPTION
**Pull request by rvarun11**
Changes examples to use [] to match the expected [Action] parameter type.

### CODE REVIEW

**Comment by mpscholten** [comment]
Will the empty list not generate a type error?

**Comment by rvarun11** [comment]
No, using the  `()` you get
```
 Couldn't match expected type ‘[Database.PostgreSQL.Simple.ToField.Action]’
```

It will get rid of that error but you may need to annotate it with the type of `model` if you get the following:
```
 The type variable ‘model0’ is ambiguous
 ```
 to fix that, you can:
 ```haskell
 (users, pagination) <- paginatedSqlQuery @User "SELECT * FROM users" []
```

Should I add this as well?

**Comment by amitaibu** [comment]
Yes, I think it's best to have a fully working example.

**Comment by rvarun11** [comment]
@amitaibu I've updated the examples as discussed :)

**Comment by mpscholten** [comment]
The type of `paginatedSqlQuery` is wrong. It needs to be a type parameter. E.g. `sqlExec` or `sqlQuery` in IHP use a generic `query` type argument with a `ToRow q` constraint:

https://github.com/digitallyinduced/ihp/blob/17f17b5f1bcc3475df2ba8e90f1c7dffd5ae9773/IHP/ModelSupport.hs#L416

**Comment by rvarun11** [comment]
Hey @mpscholten, I've updated the PR to what we discussed.

**Comment by CSchank** [comment]
@rvarun11 We played with this exact problem recently and ran into lots of weird stuff with the constraints. I believe the `FromRow` constraint is unnecessary and will constrain things more than needed. For instance, `sqlQuery` doesn't require it: https://ihp.digitallyinduced.com/api-docs/IHP-ModelSupport.html#v:sqlQuery. It would mean you can't use certain types of values as inputs, if they don't have FromRow instances, at least that's what we ran into. 

I think it seems to be needed here because of the subquery to do the count query. We actually implemented this exact function in our project a few months ago and were able to avoid the need for `FromRow`. Our code isn't open source, but I'll provide you a link to our version in case it's useful.

**Comment by CSchank** [comment]
Update: sorry, I got confused. `FromRow` is on the output result, not the input. So I think this matches what we came up with. I think for a while our type signature was more complex and I remember us running into some other weird things but we both arrived at the same type signature in the end. 😊

**Comment by rvarun11** [comment]
@mpscholten It is ready for review

File: IHP/Pagination/ControllerFunctions.hs [hunk 1]
```diff
@@ -179,12 +173,12 @@ defaultPaginationOptions =
 --
 -- *AutoRefresh:* When using 'paginatedSqlQuery' with AutoRefresh, you need to use 'trackTableRead' to let AutoRefresh know that you have accessed a certain table. Otherwise AutoRefresh will not watch table of your custom sql query.
 paginatedSqlQuery
-  :: forall model
-   . ( FromRow model
+  :: ( FromRow r
```
Comment (mpscholten) [inline IHP/Pagination/ControllerFunctions.hs:176]: Why was the type variable name changed from `model` to `r`?

File: IHP/Pagination/ControllerFunctions.hs [hunk 2]
```diff
@@ -202,14 +196,14 @@ paginatedSqlQuery = paginatedSqlQueryWithOptions defaultPaginationOptions
 --
 -- *AutoRefresh:* When using 'paginatedSqlQuery' with AutoRefresh, you need to use 'trackTableRead' to let AutoRefresh know that you have accessed a certain table. Otherwise AutoRefresh will not watch table of your custom sql query.
 paginatedSqlQueryWithOptions
-  :: forall model
-   . ( FromRow model
+  :: ( FromRow r
```
Comment (mpscholten) [inline IHP/Pagination/ControllerFunctions.hs:199]: same here

File: IHP/Pagination/ControllerFunctions.hs (hunk 1 above)
Comment (rvarun11) [reply IHP/Pagination/ControllerFunctions.hs:176]: To maintain consistency with similar functions: 

https://github.com/digitallyinduced/ihp/blob/17f17b5f1bcc3475df2ba8e90f1c7dffd5ae9773/IHP/ModelSupport.hs#L379

Comment (mpscholten) [reply IHP/Pagination/ControllerFunctions.hs:176]: Makes sense, but let's keep the old name. Single variable names are not good. (We should eventually patch `sqlQuery` to fix this there as well)

Comment (rvarun11) [reply IHP/Pagination/ControllerFunctions.hs]: I see. That makes sense. I've updated the PR accordingly.

**Review by mpscholten — APPROVED** [review]

**Pull request merged**